### PR TITLE
Improvements to dash pasing, avoid empty lists

### LINKED ISSUE #758: Failure downloading from SVTplay
Sorry to bother.

Tried (using version 1.9.6) both with the parameters I normally use
F:\##dl>svtplay-dl -q 1677 -P dash --force https://www.svtplay.se/video/16456670/nyarskonsert-fran-wien/nyarskonsert-fran-wien-1-jan-11-15-1
and got this message:
ERROR: svtplay-dl crashed
ERROR: Run again and add --verbose as an argument, to get more information
ERROR: If the error persists, you can report it at https://github.com/spaam/svtplay-dl/issues
ERROR: Include the URL used, the stack trace and the output of svtplay-dl --version in the issue

I then tried simply with special parameters
F:\##dl>svtplay-dl https://www.svtplay.se/video/16456670/nyarskonsert-fran-wien/nyarskonsert-fran-wien-1-jan-11-15-1
with the same result.

Then adding at your advice
F:\##dl>svtplay-dl --verbose https://www.svtplay.se/video/16456670/nyarskonsert-fran-wien/nyarskonsert-fran-wien-1-jan-11-15-1
which gave:
DEBUG [1514822101.6095831] site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://www.svtplay.se/video/16456670/nyarskonsert-fran-wien/nyarskonsert-fran-wien-1-jan-11-15-1'
DEBUG [1514822102.2386634] site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'http://api.svt.se/videoplayer-api/video/1383179-001A'
DEBUG [1514822102.4761963] site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-21-b.akamaized.net/se/1383179-001A/manifest.mpd'
ERROR [1514822102.6732178] site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\__init__.py/get_one_media: version: 1.9.6
Traceback (most recent call last):
  File "svtplay-dl", line 7, in <module>
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\__init__.py", line 477, in main
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\__init__.py", line 199, in get_media
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\__init__.py", line 247, in get_one_media
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\service\svtplay.py", line 76, in get
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\service\svtplay.py", line 122, in _get_video
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\fetcher\dash.py", line 122, in dashparse
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\fetcher\dash.py", line 100, in adaptionset
  File "site-packages\svtplay_dl-1.9.6-py3.5.egg\svtplay_dl\fetcher\dash.py", line 38, in templateelemt
AttributeError: 'NoneType' object has no attribute 'findall'
Failed to execute script svtplay-dl

Advice?

Thank you for being there!
/Håkan
hjmercator@gmail.com

### LINKED ISSUE #775: New problem dowloading from svtplay
Hi, I'm using Terminal and Brew on Mac to download the following link:

> https://www.svtplay.se/video/16716845/guldbaggen/guldbaggen-2018-roda-mattan-for-gala-galan-och-eftersnack

First, I got this error message:

> ERROR: svtplay-dl crashed
> ERROR: Run again and add --verbose as an argument, to get more information
> ERROR: If the error persists, you can report it at https://github.com/spaam/svtplay-dl/issues
> ERROR: Include the URL used, the stack trace and the output of svtplay-dl --version in the issue

I first upgraded svtplay-dl. When trying to dowload again I got a new error message after just pasting the link into Terminal, I didn't even have to press ENTER!

> INFO: Outfile: guldbaggen.gala.med.eftersnack-e296f59-svtplay.mp4
> Traceback (most recent call last):
>   File "/usr/local/bin/svtplay-dl", line 7, in <module>
>     svtplay_dl.main()
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 497, in main
>     get_media(urls[0], options)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 209, in get_media
>     get_one_media(stream, copy.copy(options))
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 330, in get_one_media
>     stream.download()
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/fetcher/dash.py", line 157, in download
>     self._download2(self.files)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/fetcher/dash.py", line 172, in _download2
>     eta = ETA(len(files))
> TypeError: object of type 'None

After a few seconds the following error message would follow unprompted:

> Traceback (most recent call last):
>   File "/usr/local/bin/svtplay-dl", line 7, in <module>
>     svtplay_dl.main()
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 499, in main
>     get_multiple_media(urls, options)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 182, in get_multiple_media
>     get_media(url, copy.copy(options))
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 195, in get_media
>     url, stream = generic.get(sites)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/service/__init__.py", line 115, in get
>     data = self.http.request("get", self.url).text
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/utils/__init__.py", line 68, in request
>     res = Session.request(self, method, url, verify=self.verify, proxies=self.proxy, *args, **kwargs)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/requests/sessions.py", line 508, in request
>     resp = self.send(prep, **send_kwargs)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/requests/sessions.py", line 618, in send
>     r = adapter.send(request, **kwargs)
>   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/requests/adapters.py", line 508, in send
>     raise ConnectionError(e, request=request)
> requests.exceptions.ConnectionError: HTTPConnectionPool(host='info', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x108cbfcd0>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))
> Js-MacBook-Pro:~ jhenriksson$ INFO: Outfile: guldbaggen.gala.med.eftersnack-e296f59-svtplay.mp4
> -bash: INFO:: command not found
> Js-MacBook-Pro:~ jhenriksson$ Traceback (most recent call last):
> -bash: syntax error near unexpected token `most'
> Js-MacBook-Pro:~ jhenriksson$   File "/usr/local/bin/svtplay-dl", line 7, in <module>
> -bash: syntax error near unexpected token `newline'
> Js-MacBook-Pro:~ jhenriksson$     svtplay_dl.main()
> >   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 497, in main
> -bash: syntax error near unexpected token `File'
> Js-MacBook-Pro:~ jhenriksson$     get_media(urls[0], options)
> -bash: syntax error near unexpected token `urls[0],'
> Js-MacBook-Pro:~ jhenriksson$   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 209, in get_media
> /usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py,: cannot open `/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py,' (No such file or directory)
> line:                                                                                           cannot open `line' (No such file or directory)
> 209,:                                                                                           cannot open `209,' (No such file or directory)
> in:                                                                                             cannot open `in' (No such file or directory)
> get_media:                                                                                      cannot open `get_media' (No such file or directory)
> Js-MacBook-Pro:~ jhenriksson$     get_one_media(stream, copy.copy(options))
> -bash: syntax error near unexpected token `stream,'
> Js-MacBook-Pro:~ jhenriksson$   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py", line 330, in get_one_media
> /usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py,: cannot open `/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/__init__.py,' (No such file or directory)
> line:                                                                                           cannot open `line' (No such file or directory)
> 330,:                                                                                           cannot open `330,' (No such file or directory)
> in:                                                                                             cannot open `in' (No such file or directory)
> get_one_media:                                                                                  cannot open `get_one_media' (No such file or directory)
> Js-MacBook-Pro:~ jhenriksson$     stream.download()
> >   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/fetcher/dash.py", line 157, in download
> -bash: syntax error near unexpected token `File'
> Js-MacBook-Pro:~ jhenriksson$     self._download2(self.files)
> -bash: syntax error near unexpected token `self.files'
> Js-MacBook-Pro:~ jhenriksson$   File "/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/fetcher/dash.py", line 172, in _download2
> /usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/fetcher/dash.py,: cannot open `/usr/local/Cellar/svtplay-dl/1.9.7/libexec/lib/python2.7/site-packages/svtplay_dl/fetcher/dash.py,' (No such file or directory)
> line:                                                                                               cannot open `line' (No such file or directory)
> 172,:                                                                                               cannot open `172,' (No such file or directory)
> in:                                                                                                 cannot open `in' (No such file or directory)
> _download2:                                                                                         cannot open `_download2' (No such file or directory)
> Js-MacBook-Pro:~ jhenriksson$     eta = ETA(len(files))
> -bash: syntax error near unexpected token `('
> Js-MacBook-Pro:~ jhenriksson$ TypeError: object of type 'NoneType' has no len()

I've only had time to try a few other videos from svtplay and they seem to be working. So maybe it is just this show?

Thanks in advance!

### LINKED ISSUE #810: TypeError: object of type 'NoneType' has no len()
It looks like SVT made some changes to it's 'melodifestivalen' cast as episode 2 and now 3 won't work.
I can't download it without using '-P hls' but that won't start in VLC unless I convert it.
If not using any '-P' parameters it just gives me error even with '-P dash'. Check these out. (with -v)
As the 'mello'-thing is just online for a day or so I guess I need to convert the '-P hls' version again.

```
C:\Programs\svtplay-dl>svtplay-dl.exe -v https://www.svtplay.se/video/16937270/melodifestivalen-2018/melodifestivalen-2018-deltavling-3-deltavling-3
DEBUG [1518905146.2583115] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py/get_media: version: 1.9.7
DEBUG [1518905146.2603114] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://www.svtplay.se/video/16937270/melodifestivalen-2018/melodifestivalen-2018-deltavling-3-deltavling-3'
DEBUG [1518905146.7565422] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'http://api.svt.se/videoplayer-api/video/1376650-003A'
DEBUG [1518905146.8005893] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/manifest.mpd'
DEBUG [1518905146.879218] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/master.m3u8'
DEBUG [1518905146.991678] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v3/v3.m3u8'
DEBUG [1518905147.056137] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v0/v0.m3u8'
DEBUG [1518905147.073138] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v1/v1.m3u8'
DEBUG [1518905147.091645] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v2/v2.m3u8'
DEBUG [1518905147.1098602] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v4/v4.m3u8'
DEBUG [1518905147.1252189] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v5/v5.m3u8'
DEBUG [1518905147.1408718] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v6/v6.m3u8'
DEBUG [1518905147.1579514] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/protocol_prio: Protocol priority scores (higher is better): {'hls': 4, 'http': 2, 'hds': 3, 'dash': 5, 'rtmp': 1}
INFO [1518905147.1579514] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py/get_one_media: Selected to download dash, bitrate: 2892
INFO [1518905147.1609504] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\output.py/output: Outfile: melodifestivalen.2018.avsnitt.3.deltavling.3-3bfa8c5-svtplay.mp4
Traceback (most recent call last):
  File "svtplay-dl", line 7, in <module>
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py", line 497, in main
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py", line 209, in get_media
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py", line 330, in get_one_media
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\fetcher\dash.py", line 157, in download
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\fetcher\dash.py", line 172, in _download2
TypeError: object of type 'NoneType' has no len()
[7072] Failed to execute script svtplay-dl
C:\Programs\svtplay-dl>
```

```
C:\Programs\svtplay-dl>svtplay-dl.exe -v -P dash https://www.svtplay.se/video/16937270/melodifestivalen-2018/melodifestivalen-2018-deltavling-3-deltavling-3
DEBUG [1518905221.109564] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py/get_media: version: 1.9.7
DEBUG [1518905221.112564] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://www.svtplay.se/video/16937270/melodifestivalen-2018/melodifestivalen-2018-deltavling-3-deltavling-3'
DEBUG [1518905221.6618438] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'http://api.svt.se/videoplayer-api/video/1376650-003A'
DEBUG [1518905221.698846] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/manifest.mpd'
DEBUG [1518905221.782687] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/master.m3u8'
DEBUG [1518905221.8975267] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v3/v3.m3u8'
DEBUG [1518905221.9640965] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v0/v0.m3u8'
DEBUG [1518905221.9866955] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v1/v1.m3u8'
DEBUG [1518905222.002779] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v2/v2.m3u8'
DEBUG [1518905222.0189803] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v4/v4.m3u8'
DEBUG [1518905222.0348296] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v5/v5.m3u8'
DEBUG [1518905222.0508192] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/request: HTTP getting 'https://svt-event-35-a.akamaized.net/world/1376650-003A/76228972-4f4f-44c1-abca-56cf0efee9fd/hls-v6/v6.m3u8'
DEBUG [1518905222.0647502] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\utils\__init__.py/protocol_prio: Protocol priority scores (higher is better): {'hds': 3, 'hls': 4, 'dash': 5, 'http': 2, 'rtmp': 1}
INFO [1518905222.0647502] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py/get_one_media: Selected to download dash, bitrate: 2892
INFO [1518905222.06575] site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\output.py/output: Outfile: melodifestivalen.2018.avsnitt.3.deltavling.3-3bfa8c5-svtplay.mp4
Traceback (most recent call last):
  File "svtplay-dl", line 7, in <module>
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py", line 497, in main
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py", line 209, in get_media
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\__init__.py", line 330, in get_one_media
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\fetcher\dash.py", line 157, in download
  File "site-packages\svtplay_dl-1.9.7-py3.5.egg\svtplay_dl\fetcher\dash.py", line 172, in _download2
TypeError: object of type 'NoneType' has no len()
[7000] Failed to execute script svtplay-dl

C:\Programs\svtplay-dl>
```

Gotta say that the converter doesn't really do anything, I just 'passthrough' both video and audio and after that it works.

### PR DESCRIPTION
**Pull request by dalgr**
Fixes #775
Fixes #810
Fixes #758

### CODE REVIEW

**Comment by spaam** [comment]
i have a similar fix for this that will fix for #775 but it crashes on live mpd for some reason. maybe both this and my fix both 🤔

**Comment by dalgr** [comment]
What source did you try for the live mpd?

**Comment by dalgr** [comment]
@spaam 
I have added a fix for MPD live

**Comment by dalgr** [comment]
@spaam 
I have updated with a fix for melodifestivalen
Tested with:
```
svtplay-dl "https://www.svtplay.se/video/17048166/melodifestivalen-2018/melodifestivalen-2018-deltavling-4-deltavling-4" -f -v
```

**Comment by spaam** [comment]
oh nice :)

File: lib/svtplay_dl/fetcher/dash.py [hunk 1]
```diff
@@ -25,43 +25,60 @@ def __init__(self, url):
...
+    duration = 1
+    total = 1
 
     init = element.attrib["initialization"]
     media = element.attrib["media"]
     if "startNumber" in element.attrib:
         start = int(element.attrib["startNumber"])
     else:
-        start = 0
+        start = 1
```
Comment (spaam) [inline lib/svtplay_dl/fetcher/dash.py]: why change this number to `1`?

Comment (dalgr) [reply lib/svtplay_dl/fetcher/dash.py]: I noticed that the Default in MPD files usually was 1.

**Comment by dalgr** [comment]
@spaam 
Time for a new release when this if fixed?

File: lib/svtplay_dl/fetcher/dash.py (hunk 1 above)
Comment (spaam) [reply lib/svtplay_dl/fetcher/dash.py]: ahh true. :)

**Comment by spaam** [comment]
@dalgr  yes.  after the fix for dplay channel streaming that im working on now.

**Pull request merged**